libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp

```diff
@@ -30,6 +31,110 @@ bool AP_BattMonitor_SMBus::get_cycle_count(uint16_t &cycles) const
     return true;
 }
 
+// return true if serial number can be provided and fills in serial number argument
+bool AP_BattMonitor_SMBus::get_serial_number(char *serial_number, uint8_t buflen) const {
+
+    if (_serial_number != -1) {
+        snprintf(serial_number, buflen, "%i", (signed) _serial_number);
+        return true;
+    }
+
+    return false;
+}
+
+// return true if device name can be provided and fills in device name
+bool AP_BattMonitor_SMBus::read_device_name()
+{
+    if (_device_name_len != 0) {
+        return true;
+    }
+
+    _device_name_len = read_block(BATTMONITOR_SMBUS_DEVICE_NAME, _device_name, true) + 1;
+
+    return _device_name_len > 0;
+}
+
+// return true if device name can be provided and fills in product name
+bool AP_BattMonitor_SMBus::get_device_name(char * device_name, uint8_t buflen) const
+{ 
+    // check device name
+    if (_device_name_len) {
+        strncpy(device_name, (char*)_device_name, buflen -1);
+        return true;
+    }
+
+    return false;
+}
+
+// return true if manufacturer name can be provided and fills in device name
+bool AP_BattMonitor_SMBus::read_manufacturer_name()
+{
+    // already have manufacturer name return
+    if (_manufacturer_name_len != 0) {
+        return true;
+    }
+
+    _manufacturer_name_len = read_block(BATTMONITOR_SMBUS_MANUFACTURE_NAME, _manufacturer_name, true) + 1;
+
+    return _manufacturer_name_len > 0;
+}
+
+// return true if manufacturer name can be provided and fills in manufacturer name
+bool AP_BattMonitor_SMBus::get_manufacturer_name(char * manufacturer_name, uint8_t buflen) const
+{    
+    if (_manufacturer_name_len) {
+        strncpy(manufacturer_name, (char*)_manufacturer_name, buflen -1);
```

Comment:
You need to manually null terminate the array here, it could have been not initialized on the stack, which can lead to later reading of garbage memory. (This is the actual use case here). Also why would we bother copying more then _manufacturer_name_len of bytes? If we are storing the length why aren't we using it?

Reply:
Fixed it to copy the correct # of bytes and null terminate the last byte.